c2rust-analyze/src/rewrite/expr/distribute.rs

```diff
@@ -0,0 +1,126 @@
+    tcx: TyCtxt<'tcx>,
+    unlower_map: BTreeMap<(Location, Vec<SubLoc>), MirOrigin>,
+    mir_rewrites: HashMap<Location, Vec<MirRewrite>>,
+) -> HashMap<HirId, Vec<mir_op::RewriteKind>> {
+    let mut info_map = HashMap::<HirId, Vec<RewriteInfo>>::new();
+
+    for (loc, mir_rws) in mir_rewrites {
+        for mir_rw in mir_rws {
+            let mut key = (loc, mir_rw.sub_loc);
+
+            let mut origin = unlower_map.get(&key);
+            if origin.is_none() && matches!(key.1.last(), Some(&SubLoc::RvalueOperand(0))) {
+                // Hack: try without the `RvalueOperand(0)` at the end.
+                // TODO: remove SubLoc::RvalueOperand from mir_op
+                key.1.pop();
+                origin = unlower_map.get(&key);
+                if origin.is_none() {
+                    key.1.push(SubLoc::RvalueOperand(0));
+                }
+            }
+            let mut origin = match origin {
+                Some(x) => x,
+                None => {
+                    error!("unlower_map has no origin for {:?}", key);
+                    continue;
+                }
+            };
+
+            if origin.desc == MirOriginDesc::StoreIntoLocal && key.1.len() == 0 {
+                // Hack: try with an extra `Rvalue` in the key.
+                // TODO: add SubLoc::Rvalue in mir_op ptr::offset handling
+                key.1.push(SubLoc::Rvalue);
+                match unlower_map.get(&key) {
+                    Some(o) if o.desc == MirOriginDesc::Expr => {
+                        origin = o;
+                    }
+                    _ => {
+                        key.1.pop();
+                    }
+                }
+            }
+
+            if origin.desc != MirOriginDesc::Expr {
+                error!(
+                    "can't distribute rewrites onto {:?} origin {:?}\n\
+                        key = {:?}\n\
+                        mir_rw = {:?}",
+                    origin.desc, origin, key, mir_rw.kind
+                );
+            }
+
+            info_map
+                .entry(origin.hir_id)
+                .or_default()
+                .push(RewriteInfo {
+                    rw: mir_rw.kind,
+                    loc,
+                    sub_loc: key.1,
+                    desc: origin.desc,
+                });
+        }
+    }
+
+    // If a single `HirId` has rewrites from multiple different pieces of MIR, it's ambiguous how
+    // to order those rewrites.  (`mir_rewrites` only establishes an ordering between rewrites on
+    // the same `Location`.)  For now, we complain if we see this ambiguity; in the future, we may
+    // need to add rules to resolve it in a particular way, such as prioritizing one `SubLoc` or
+    // `MirOriginDesc` over another.
+    for (&hir_id, infos) in &info_map {
+        let all_same_loc = infos.iter().skip(1).all(|i| {
+            i.loc == infos[0].loc && i.sub_loc == infos[0].sub_loc && i.desc == infos[0].desc
+        });
```

Comment:
If we add `itertools` as a dependency, which we already have in most other `c2rust-*` crates, we can use their `.all_equal()` method.  And I think the `.map(|i| (i.loc, &i.sub_loc, i.desc))` more clearly shows which fields are being checked.

```suggestion
        let all_same_loc = infos
            .iter()
            .map(|i| (i.loc, &i.sub_loc, i.desc))
            .all_equal();
```

Reply:
Done.